tmpl/blog/mirage-entropy.md

```diff
@@ -0,0 +1,362 @@
+much we try, and in the extreme, this means that it cannot be generated by a
+program. While the latter restriction is obviously a little too strong for our
+purpose, it highlights a deep distinction in what people mean by being "random".
+Jumping around is one thing. Being actually unpredictable is a wholly different
+matter.
+
+There are many other [simple][wolfram-rule-30] mathematical processes which
+generate sequences with high statistical randomness. Many of those are used to
+produce "random" sequences of various purposes. But these are still completely
+deterministic processes that exhibit random behaviour only in the statistical
+sense. Instead of being *random*, they are *pseudo-random*, and we call such
+generators Pseudo-Random Number Generators (PRNGs).
+
+We can look for something approaching a "true" random sequence in the nature.
+The current agreement is that the nature of quantum processes is random in this
+sense, and random sequences based on this idea are readily available for
+[download][dgrb-boskovic]. Or we can use the microphone and keep recording; the
+lowest-order bits of the signal are pretty unpredictable. But we cannot write a
+program to generate an actually random sequence.
+
+Still, we need to compromise. The real problem of common PRNGs is that knowing
+the rule and observing some of the outputs is enough to predict the rest of the
+sequence. The entire future behavior of [Mersenne twister][mersenne-twister],
+one of the most commonly used generators in various programming packages, can be
+predicted after observing only 624 outputs in a row. A step up from such a
+process is a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG).
+Their key property that it is computationally prohibitively expensive to
+distinguish their outputs from a "true" random sequence. This also means that it
+is computationally prohibitively expensive to reconstruct their internal state,
+just by looking at their outputs. In a sense, someone trying to predict the
+outputs can not take shortcuts, and is instead forced to perform the laborious
+task of starting the generator with all the possible states and checking if the
+output matches the observed sequence. This is how we can quantify a CSPRNG
+unpredictability: it takes trying about half of all the possibilities to guess
+the state.
+
+MirageOS's security stack contains a CSPRNG, a design called [Fortuna][fortuna].
+What it really does, is encrypt the simple sequence `0, 1, 2, 3, ...` with AES
+(AES-CTR) using a secret key. This makes it as resistant to prediction as AES is
+to [known-plaintext attacks][wiki-known-plaintext]. After each output, it
+generates a bit more, hashes that, and uses the result as the next key. This is
+not to improve the statistical randomness, as it is already guaranteed by AES.
+Rather, it's a form of [forward secrecy][wiki-forward-secrecy]: an attacker who
+learns the secret key at some point would need to perform the [preimage
```

Comment:
preimage or pre-image ?
(I prefer latter but if former is standard, stick with it.)